src/sentry/snuba/discover.py

```diff
@@ -0,0 +1,207 @@
+        if isinstance(cond[0], (list, tuple)):
+            if get_function_index(cond[0]) is not None:
+                cond[0] = resolve_condition(cond[0])
+                return cond
+            else:
+                # Nested conditions
+                return [resolve_condition(item) for item in cond]
+    raise ValueError("Unexpected condition format %s" % cond)
+
+
+def resolve_discover_aliases(snuba_args):
+    """
+    Resolve the public schema aliases to the discover dataset.
+
+    Returns a copy of the input structure, and includes a
+    `translated_columns` key containing the selected fields that need to
+    be renamed in the result set.
+    """
+    resolved = deepcopy(snuba_args)
+    translated_columns = {}
+    derived_columns = set()
+
+    selected_columns = resolved.get("selected_columns")
+    if selected_columns:
+        for (idx, col) in enumerate(selected_columns):
+            if isinstance(col, (list, tuple)):
+                raise ValueError("discover selected_columns should only be str. got %s" % col)
+            name = resolve_column(col)
+            selected_columns[idx] = name
+            translated_columns[name] = col
+        resolved["selected_columns"] = selected_columns
+
+    groupby = resolved.get("groupby")
+    if groupby:
+        for (idx, col) in enumerate(groupby):
+            name = col
+            if col not in derived_columns:
+                name = resolve_column(col)
+            groupby[idx] = name
+        resolved["groupby"] = groupby
+
+    aggregations = resolved.get("aggregations")
+    for aggregation in aggregations or []:
+        derived_columns.add(aggregation[2])
+        if isinstance(aggregation[1], six.string_types):
+            aggregation[1] = resolve_column(aggregation[1])
+        elif isinstance(aggregation[1], (set, tuple, list)):
+            aggregation[1] = [resolve_column(col) for col in aggregation[1]]
+    resolved["aggregations"] = aggregations
+
+    conditions = resolved.get("conditions")
+    if conditions:
+        for (i, condition) in enumerate(conditions):
+            replacement = resolve_condition(condition)
+            conditions[i] = replacement
+        resolved["conditions"] = list(filter(None, conditions))
+
+    # TODO add support for extracting having conditions.
+
+    orderby = resolved.get("orderby")
+    if orderby:
+        orderby = orderby if isinstance(orderby, (list, tuple)) else [orderby]
+        resolved_orderby = []
+
+        for field_with_order in orderby:
+            field = field_with_order.lstrip("-")
+            resolved_orderby.append(
+                u"{}{}".format(
+                    "-" if field_with_order.startswith("-") else "",
+                    field if field in derived_columns else resolve_column(field),
+                )
+            )
+        resolved["orderby"] = resolved_orderby
+    return resolved, translated_columns
+
+
+def query(selected_columns, query, params, orderby=None, referrer=None):
```

Comment:
This seems very similar to the `transform_aliases_and_query` method except this is hardcoded to the Discover dataset and that method is hardcoded to Events (correct me if this is not a correct analysis). What do think about having some shared methods that accept a Dataset parameter and can do the transformation for the requested dataset. It could be used by both this method as well as `transform_aliases_and_query`. Otherwise we will have 2 bits of code that are doing the same complicated processing just for different datasets.

Reply:
The `transform_aliases_and_query` method currently uses `dataset_query` and can be pinned to the events dataset but we don't use that feature anywhere currently.

My plan was to adapt discover1 to use this new abstraction and entirely remove `transform_aliases_and_query` once these new methods stabilize. While that does enable customers to use transaction fields in discover1 it also gets us to a more consistent public schema faster. With the discover1 UI not showing transaction fields it should be hard for customers to find out that old discover supports them.

Reply:
Ok. If we make sure to append the condition ["type", "!=", "transaction"] in the API when we do the conversion to use this code, it'll make sure that the Events dataset is always selected, regardless of what non surfaced fields the user might find.